File: python/cfar.py
"""
CA-CFAR (Cell-Averaging Constant False Alarm Rate) Detector.

Adaptive threshold detection that estimates the local noise floor
from training cells surrounding each cell-under-test (CUT).
"""

import numpy as np
import matplotlib.pyplot as plt

# Default CFAR parameters
DEFAULT_NUM_TRAIN = 64     # Training cells on each side of the CUT
DEFAULT_NUM_GUARD = 32      # Guard cells on each side of the CUT
DEFAULT_PFA = 1e-3         # Desired probability of false alarm


def ca_cfar(magnitudes, num_train, num_guard, pfa):
    """
    Run CA-CFAR detection on a 1D magnitude array.

    For each cell-under-test (CUT):
      1. Average the training cells on both sides (skip guard cells)
      2. Multiply by a threshold factor derived from pfa
      3. If CUT exceeds the threshold, flag as detection

    Args:
        magnitudes: 1D numpy array of real values (e.g. |matched_filter_output|^2)
        num_train:  Number of training cells on EACH side of the CUT
        num_guard:  Number of guard cells on EACH side of the CUT
        pfa:        Desired probability of false alarm

    Returns:
        detections: Boolean numpy array, True where a target is detected
        thresholds: Numpy array of the adaptive threshold at each cell
    """
    # Initialize detection array and threshold array
    num_samples = len(magnitudes)
    detections = np.zeros(num_samples, dtype=bool)
    thresholds = np.zeros(num_samples)
    
    # The total number of training cells used for threshold calculation
    N_train_total = 2 * num_train
    
    # Calculate the constant threshold factor once
    alpha = compute_threshold_factor(2*num_train, pfa)

    for i in range(num_samples):
        # Skip guard cells (indices from num_guard to num_guard-1, and num_samples-num_guard to num_samples-1)
        if (i < num_guard) or (i >= num_samples - num_guard):
            thresholds[i] = 0 # Set threshold to 0 or some sentinel value for guard cells
            continue

        # Define the window boundaries for training cells
        start_train = max(0, i - num_train - num_guard)
        end_train = min(num_samples - 1, i + num_train + num_guard)
        
        # The actual training window is [i - num_train - num_guard, i - num_guard - 1]
        # and [i + num_guard + 1, i + num_train + num_guard]
        
        # Left training window: [i - num_train - num_guard, i - num_guard - 1]
        left_start = max(0, i - num_train - num_guard)
        left_end = i - num_guard - 1
        
        # Right training window: [i + num_guard + 1, i + num_train + num_guard]
        right_start = i + num_guard + 1
        right_end = min(num_samples - 1, i + num_train + num_guard)
        
        # Extract training cell magnitudes
        left_train = magnitudes[left_start:left_end + 1]
        right_train = magnitudes[right_start:right_end + 1]
        
        # Average the training cells (assuming equal weighting)
        # We must handle cases where the window is truncated at the edges
        avg_train = np.mean(np.concatenate((left_train, right_train)))
        
        # Calculate the adaptive threshold
        thresholds[i] = alpha * avg_train
        
        # Detection decision
        if magnitudes[i] > thresholds[i]:
            detections[i] = True
            
    return detections, thresholds


def compute_threshold_factor(num_train, pfa):
    """
    Compute the CFAR threshold multiplier from the number of
    training cells and desired probability of false alarm.

    Args:
        num_train: Total number of training cells (both sides combined)
        pfa:       Desired probability of false alarm

    Returns:
        alpha: Threshold scaling factor
    """
    # The threshold factor alpha is derived from the chi-squared distribution
    # for the sum of squared normalized Gaussian random variables.
    # For a given PFA and N_train, alpha is often approximated or calculated
    # using the inverse CDF of the chi-squared distribution.
    # A common approximation for the threshold factor is:
    # alpha = (N_train * (1 - pfa))^(-1/2) * (1 + 2/N_train)
    # However, for simplicity and matching typical DSP textbook examples,
    # we use a simplified form derived from the assumption of Gaussian noise.
    
    # A more robust approach often involves the inverse CDF of chi-squared.
    # For this implementation, we will use a common simplified form:
    return num_train *(pfa ** (-1/num_train) -1)


def plot_cfar(magnitudes, thresholds, detections, sample_rate):
    """
    Plot the magnitude signal, adaptive threshold, and detection flags.

    Args:
        magnitudes: Input magnitude array
        thresholds: Adaptive threshold from ca_cfar()
        detections: Boolean detection array from ca_cfar()
        sample_rate: Sample rate (Hz)
    """
    # Time vector for plotting
    time_vector = np.arange(len(magnitudes)) / sample_rate

    plt.figure(figsize=(14, 10))

    # Plot 1: Magnitude Signal
    plt.subplot(3, 1, 1)
    plt.plot(time_vector, magnitudes, label='Magnitude $|R|^2$')
    plt.title('Raw Magnitude Signal')
    plt.xlabel('Time (s)')
    plt.ylabel('Magnitude')
    plt.grid(True)
    plt.legend()

    # Plot 2: Adaptive Threshold
    plt.subplot(3, 1, 2)
    plt.plot(time_vector, thresholds, label='Adaptive Threshold $\\alpha \\cdot \\text{AvgTrain}$', color='red')
    plt.title('Adaptive Threshold')
    plt.xlabel('Time (s)')
    plt.ylabel('Threshold Value')
    plt.grid(True)
    plt.legend()

    # Plot 3: Detection Results
    plt.subplot(3, 1, 3)
    # Plot the signal again for context
    plt.plot(time_vector, magnitudes, label='Magnitude $|R|^2$', alpha=0.5)
    # Plot the detection flags (True=Detection)
    plt.plot(time_vector[detections], magnitudes[detections], 'ro', markersize=5, label='Detection')
    plt.plot(time_vector, thresholds, 'r--', label='Threshold')
    plt.title('Detection Results')
    plt.xlabel('Time (s)')
    plt.ylabel('Magnitude')
    plt.grid(True)
    plt.legend()

    plt.tight_layout()
    plt.show()


if __name__ == "__main__":
    # Test with synthetic data: a few peaks embedded in noise
    pass

File: java/Cfar.java
/*
 * ===============================================================================
 * Cfar — Cell-Averaging Constant False Alarm Rate detector.
 * ===============================================================================
 *
 * Given a magnitude array (typically |matched_filter_output|²), decide
 * which cells are targets and which are noise. The threshold for each
 * cell is computed adaptively from the surrounding cells' noise level,
 * so the false-alarm rate stays roughly constant regardless of the
 * absolute noise floor.
 *
 * Window layout around each cell-under-test:
 *
 *   [ TRAIN | GUARD | CUT | GUARD | TRAIN ]
 *     ^       ^       ^     ^       ^
 *     |       |       |     |       \
 *     |       |       |     \        right training cells
 *     |       |       |      right guard cells
 *     |       |       cell under test (the one we're deciding about)
 *     |       left guard cells (skipped to avoid target-self contamination)
 *     left training cells (used to estimate local noise floor)
 *
 * Training cells: big enough to get a stable noise estimate.
 * Guard cells:   big enough to keep the target's own energy (which
 *                spreads across many samples after the matched filter)
 *                from contaminating the training cells.
 * ===============================================================================
 */
public class Cfar {

    /**
     * Compute the CFAR threshold scaling factor alpha.
     *
     * Derivation (sketch):
     *   Under a "noise only" hypothesis, if the matched filter output is
     *   complex Gaussian, then |MF|² follows an exponential distribution.
     *   The sum of N training cells (each exponential) divided by N is a
     *   chi-squared-ish quantity whose distribution is known. We set the
     *   probability that |MF(CUT)|² exceeds alpha * (training average) to
     *   equal the desired false-alarm probability PFA, solve for alpha,
     *   and get:
     *
     *       alpha = N * (PFA^(-1/N) - 1)
     *
     *   where N is the total number of training cells (both sides combined).
     *   Higher N gives a more stable estimate; lower PFA pushes alpha up
     *   (fewer false alarms but potentially missing weak targets).
     *
     *   You don't need to re-derive this in an interview, but you should be
     *   able to say: "alpha comes from solving the false-alarm probability
     *   for a known noise distribution; it scales with 1/PFA and depends on
     *   how many training cells we use for the noise estimate."
     */
    public static double computeThresholdFactor(int totalTrainingCells,
                                                double probabilityFalseAlarm) {
        return totalTrainingCells
             * (Math.pow(probabilityFalseAlarm, -1.0 / totalTrainingCells) - 1.0);
    }

    /**
     * Sliding-window CA-CFAR.
     *
     * For each sample in `magnitudes`:
     *   1. Treat that sample as the "cell under test" (CUT).
     *   2. Look at the training cells on both sides (skipping the nearby
     *      guard cells that could be contaminated by the CUT's own energy).
     *   3. Average those training cells to estimate the local noise floor.
     *   4. Multiply the average by alpha (from computeThresholdFactor).
     *   5. If the CUT exceeds that threshold, flag a detection.
     *
     * The "constant false alarm rate" property comes from the fact that the
     * threshold always sits at the same statistical distance above the local
     * noise — no matter what the absolute noise level is, the rate of
     * false alarms per cell stays roughly constant.
     *
     * Edge handling: cells near the beginning or end of the array don't have
     * a complete window around them. The simplest approach — used here — is
     * to skip those cells entirely (no detection, threshold = 0).
     *
     * The `thresholdsOut` parameter is filled in as a side effect, purely
     * so the caller can plot the adaptive threshold alongside the data.
     */
    public static boolean[] detect(double[] magnitudes,
                                   int trainingCellsPerSide,
                                   int guardCellsPerSide,
                                   double probabilityFalseAlarm,
                                   double[] thresholdsOut) {
        int numSamples = magnitudes.length;
        boolean[] detections = new boolean[numSamples];

        // Compute alpha once, reuse for every cell.
        double thresholdScalingFactor = computeThresholdFactor(
                2 * trainingCellsPerSide, probabilityFalseAlarm);

        // Half-window size from the CUT out to the edge of the training
        // region on one side. This is the minimum distance from the edge
        // of the array for which we have a full window.
        int halfWindowSize = trainingCellsPerSide + guardCellsPerSide;

        for (int cellUnderTest = 0; cellUnderTest < numSamples; cellUnderTest++) {
            // Skip edge cells without a full window.
            if (cellUnderTest < halfWindowSize
             || cellUnderTest >= numSamples - halfWindowSize) {
                thresholdsOut[cellUnderTest] = 0.0;
                continue;
            }

            // Sum the training cells on both sides, skipping guard cells.
            double trainingSum = 0.0;

            // Left side: from (CUT - halfWindow) up to (CUT - guard - 1).
            for (int trainingCell = cellUnderTest - halfWindowSize;
                 trainingCell <= cellUnderTest - guardCellsPerSide - 1;
                 trainingCell++) {
                trainingSum += magnitudes[trainingCell];
            }
            // Right side: from (CUT + guard + 1) up to (CUT + halfWindow).
            for (int trainingCell = cellUnderTest + guardCellsPerSide + 1;
                 trainingCell <= cellUnderTest + halfWindowSize;
                 trainingCell++) {
                trainingSum += magnitudes[trainingCell];
            }

            // Average over the total number of training cells.
            double trainingAverage = trainingSum / (2 * trainingCellsPerSide);

            // Scale by alpha to get the detection threshold for this cell.
            double adaptiveThreshold = thresholdScalingFactor * trainingAverage;
            thresholdsOut[cellUnderTest] = adaptiveThreshold;

            // Detection decision.
            if (magnitudes[cellUnderTest] > adaptiveThreshold) {
                detections[cellUnderTest] = true;
            }
        }
        return detections;
    }
}
